File: Code/TEST_python/heap_sort.py
# Credits: https://www.programiz.com/dsa/heap-sort

# Heap Sort in python


def heapify(arr, n, i):
    # Find largest among root and children
    largest = i
    l = 2 * i + 1
    r = 2 * i + 2

    if l < n and arr[i] < arr[l]:
        largest = l

    if r < n and arr[largest] < arr[r]:
        largest = r

    # If root is not largest, swap with largest and continue heapifying
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)


def heapSort(arr):
    n = len(arr)

    # Build max heap
    for i in range(n // 2, -1, -1):
        heapify(arr, n, i)

    for i in range(n - 1, 0, -1):
        # Swap
        arr[i], arr[0] = arr[0], arr[i]

        # Heapify root element
        heapify(arr, i, 0)


with open("heap_sort_data.txt") as f:
    arr = list(map(int, f.readline().split()))
heapSort(arr)
print("Sorted")

File: Code/TEST_c/heap_sort.c
// Heap Sort in C
  
  #include <stdio.h>
  #include <stdlib.h>

  // Function to swap the the position of two elements
  void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
  }
  
  void heapify(int arr[], int n, int i) {
    // Find largest among root, left child and right child
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;
  
    if (left < n && arr[left] > arr[largest])
      largest = left;
  
    if (right < n && arr[right] > arr[largest])
      largest = right;
  
    // Swap and continue heapifying if root is not largest
    if (largest != i) {
      swap(&arr[i], &arr[largest]);
      heapify(arr, n, largest);
    }
  }
  
  // Main function to do heap sort
  void heapSort(int arr[], int n) {
    // Build max heap
    for (int i = n / 2 - 1; i >= 0; i--)
      heapify(arr, n, i);
  
    // Heap sort
    for (int i = n - 1; i >= 0; i--) {
      swap(&arr[0], &arr[i]);
  
      // Heapify root element to get highest element at root again
      heapify(arr, i, 0);
    }
  }
  
  // Print an array
  void printArray(int arr[], int n) {
    for (int i = 0; i < n; ++i)
      printf("%d ", arr[i]);
    printf("\n");
  }
  
  // Driver code
  int main() {

    FILE *inFile = fopen("heap_sort_data.txt", "r");

    if (!inFile) {
        fprintf(stderr, "Error opening file\n");
        return 1;
    }

    const int numValues = 10000000;
    int *arr = malloc(numValues * sizeof(int));
    int val;
 
    // Read values from first row into arr array
    for (int i = 0; i < numValues; i++) {
        fscanf(inFile, "%d", &val);
        arr[i] = val;
    }

    fclose(inFile);

    int n = sizeof(arr) / sizeof(arr[0]);
  
    heapSort(arr, n);

    printf("Sorted");

    return 0;
  }